pkg/deployment/resources/servicemonitor.go

```diff
@@ -0,0 +1,118 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Copyright holder is ArangoDB GmbH, Cologne, Germany
+//
+// Author Max Neunhoeffer
+//
+
+package resources
+
+import (
+	"github.com/arangodb/kube-arangodb/pkg/util/k8sutil"
+
+	coreosv1 "github.com/coreos/prometheus-operator/pkg/apis/monitoring/v1"
+	clientv1 "github.com/coreos/prometheus-operator/pkg/client/versioned/typed/monitoring/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+)
+
+func LabelsForExporterServiceMonitor(deploymentName string) map[string]string {
+	return map[string]string{
+		k8sutil.LabelKeyArangoDeployment: deploymentName,
+		k8sutil.LabelKeyApp:              "arango-exporter",
+		"context":                        "metrics",
+	}
+}
+
+func LabelsForExporterServiceMonitorSelector(deploymentName string) map[string]string {
+	return map[string]string{
+		k8sutil.LabelKeyArangoDeployment: deploymentName,
+		k8sutil.LabelKeyApp:              "arangodb",
+	}
+}
+
+// EnsureServiceMonitor creates or updates a ServiceMonitor.
+func (r *Resources) EnsureServiceMonitor() error {
+	// Some preparations:
+	log := r.log
+	apiObject := r.context.GetAPIObject()
+	deploymentName := apiObject.GetName()
+	ns := apiObject.GetNamespace()
+	owner := apiObject.AsOwner()
+	spec := r.context.GetSpec()
+	if !spec.Metrics.IsEnabled() {
+		return nil
+	}
+	serviceMonitorName := deploymentName + "-exporter"
+
+	// First get a client:
+	var restConfig *rest.Config
+	restConfig, err := k8sutil.InClusterConfig()
+	if err != nil {
+		return maskAny(err)
+	}
+	var mClient *clientv1.MonitoringV1Client
+	mClient, err = clientv1.NewForConfig(restConfig)
+	if err != nil {
+		return maskAny(err)
+	}
+
+	// Check if ServiceMonitor already exists
+	serviceMonitors := mClient.ServiceMonitors(ns)
+	_, err = serviceMonitors.Get(serviceMonitorName, metav1.GetOptions{})
+	if err != nil {
+		if k8sutil.IsNotFound(err) {
+			// Need to create one:
+			smon := &coreosv1.ServiceMonitor{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:            serviceMonitorName,
+					Labels:          LabelsForExporterServiceMonitor(deploymentName),
+					OwnerReferences: []metav1.OwnerReference{owner},
+				},
+				Spec: coreosv1.ServiceMonitorSpec{
+					JobLabel: "k8s-app",
+					Endpoints: []coreosv1.Endpoint{
+						coreosv1.Endpoint{
+							Port:     "exporter",
+							Interval: "10s",
+							Scheme:   "https",
+							TLSConfig: &coreosv1.TLSConfig{
+								InsecureSkipVerify: true,
+							},
```

Comment:
You had this brilliant idea to add the cluster ca here :+1:

Reply:
Indeed, now I have to deliver... :-(

Reply:
And it might also not always be TLS...

Reply:
Found no way to actually hand in the cert. Go with insecure for now.